core/src/layers/madsim.rs

```diff
@@ -0,0 +1,318 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::ops::{OpList, OpRead, OpScan, OpWrite};
+use crate::raw::oio::Entry;
+use crate::raw::{oio, Accessor, Layer, LayeredAccessor, RpList, RpRead, RpScan, RpWrite};
+use crate::{EntryMode, Metadata};
+use async_compat::CompatExt;
+use async_trait::async_trait;
+use bytes::Bytes;
+use madsim::plugin::node;
+use madsim::plugin::simulator;
+use madsim::plugin::Simulator;
+use madsim::rand::GlobalRng;
+use madsim::task::NodeId;
+use madsim::time::TimeHandle;
+use madsim::Config;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::SeekFrom;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+#[derive(Debug, Copy, Clone)]
+pub struct MadsimLayer {
+    service_node_id: NodeId,
+}
+
+impl MadsimLayer {
+    pub fn new(service_node_id: NodeId) -> Self {
+        Self { service_node_id }
+    }
+}
+
+impl<A: Accessor> Layer<A> for MadsimLayer {
+    type LayeredAccessor = MadsimAccessor;
+
+    fn layer(&self, _inner: A) -> Self::LayeredAccessor {
+        let handle = simulator::<ServiceSim>().get_node(self.service_node_id);
+        MadsimAccessor { handle }
+    }
+}
+
+#[derive(Debug)]
+pub struct MadsimAccessor {
+    handle: ServiceNodeHandle,
+}
+
+#[async_trait]
+impl LayeredAccessor for MadsimAccessor {
+    type Inner = ();
+    type Reader = MadsimReader;
+    type BlockingReader = MadsimReader;
+    type Writer = MadsimWriter;
+    type BlockingWriter = MadsimWriter;
+    type Pager = MadsimPager;
+    type BlockingPager = MadsimPager;
+
+    fn inner(&self) -> &Self::Inner {
+        &()
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> crate::Result<(RpRead, Self::Reader)> {
+        let handle = self.handle.inner.lock().unwrap();
+        let length = handle
+            .get(path)
+            .map(|(_, data)| data.len())
+            .unwrap_or_default();
+        Ok((
+            RpRead::new(length as u64),
+            MadsimReader {
+                path: path.to_string(),
+                args,
+                handle: self.handle.clone(),
+            },
+        ))
+    }
+
+    async fn write(&self, path: &str, args: OpWrite) -> crate::Result<(RpWrite, Self::Writer)> {
+        Ok((
+            RpWrite::new(),
+            MadsimWriter {
+                path: path.to_string(),
+                args,
+                handle: self.handle.clone(),
+            },
+        ))
+    }
+
+    async fn list(&self, path: &str, args: OpList) -> crate::Result<(RpList, Self::Pager)> {
+        todo!()
```

Comment:
Please forward to inner instead of `todo!()`

Reply:
Also, we can return an unsupported errors here. And implement this feature in the future.